qiskit/transpiler/passmanager.py

```diff
@@ -133,9 +165,23 @@ def run(self, circuit):
         del circuit
         self.reset()  # Reset passmanager instance before starting
 
+        count = 0
         for passset in self.working_list:
             for pass_ in passset:
+                if self.callback:
+                    start_time = time()
                 dag = self._do_pass(pass_, dag, passset.options)
+                if self.callback:
+                    end_time = time()
```

Comment:
timing the `_do_pass` call won't be accurate because this function does a pass as well as any requirements of that pass, which may be other passes. 

I think it has to use the context around the call to `_run_this_pass()` which @1ucian0 suggested (line 221 of this file)

Reply:
Right, I saw how the time was being done there, but it raised larger questions about how we not just measure time but also when we call the callback and what the expectation is for what's passed to it. If we get the time like where the context manager does now that can result in multiple times being generated per iteration in the `passmanager.run()` loop, but with only 1 callback function call. So the question is what do we consider a pass for the callback function is it the full pass (including the execution of any requirements) or each inidivual Pass object that we call `run()` on. I did time this way under the assumption it was the former and we were going to treat the complete execution of `_do_pass()` as what the callback deals with, and used that for the timing. If we want it to be each call of run on a pass object I can change that, but we can't really do it both ways.

Reply:
I think the debugger should explicitly show all the passes that were run, because that is useful information that can help debugging. For example say I create a pass manager with a single pass like this and debug it during run:

```python
from qiskit.transpiler import PassManager
from qiskit.transpiler.passes import CommutativeCancellation

def callback(pass_, dag, time, property_set, count):
     print(pass_)

pm = PassManager()
pm.append(CommutativeCancellation())
pm.callback = callback

circ = QuantumCircuit(3)
circ.z(0)
circ.cx(0, 2)
circ.z(0)

pm.run(circ)
```
It will show exactly one pass as being run, even though really two passes are run (`CommutationAnalysis`, then, `CommutiveCancellation`), because of the `requires` relationship between the passes which the PassManager fulfills.

Reply:
Ok I reworked this to call the callback right after each `pass_.run()` call. To do it cleanly required removing the use of the context manager because it got tricky to extract the time measurements from it and adding the callback to the context manager made it overly complex.